File: DebateUser/DebateUserBackEnd/src/main/java/com/debate/service/UserService.java
package com.debate.service;

import com.debate.dto.response.UserRankingResponse;
import com.debate.dto.response.UserResponse;
import com.debate.entity.Debate;
import com.debate.entity.User;
import com.debate.exception.ResourceNotFoundException;
import com.debate.repository.DebateOpinionRepository;
import com.debate.repository.DebateRepository;
import com.debate.repository.CommentRepository;
import com.debate.repository.LikeRepository;
import com.debate.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;
import org.springframework.data.domain.Sort;

@Service
@RequiredArgsConstructor
public class UserService {
    private final UserRepository userRepository;
    private final DebateRepository debateRepository;
    private final CommentRepository commentRepository;
    private final LikeRepository likeRepository;
    private final DebateOpinionRepository debateOpinionRepository;

    /**
     * 받은 좋아요 수 기준 사용자 랭킹 조회
     *
     * @param limit 조회할 상위 사용자 수
     * @return 사용자 랭킹 목록
     */
    /**
     * 사용자 랭킹 조회 (기간 및 기준별)
     *
     * @param period 기간 (daily, monthly, yearly, all)
     * @param criteria 기준 (likes, votes, comments)
     * @param limit 조회할 상위 사용자 수
     * @return 사용자 랭킹 목록
     */
    public List<UserRankingResponse> getUserRanking(String period, String criteria, int limit) {
        java.time.LocalDateTime end = java.time.LocalDateTime.now();
        java.time.LocalDateTime start = calculateStartDate(period, end);
        Pageable pageable = org.springframework.data.domain.PageRequest.of(0, limit);

        List<UserRankingResponse> ranking;

        switch (criteria.toLowerCase()) {
            case "votes": // 투표율 (의견 수)
                ranking = userRepository.findRankingByDebateOpinions(start, end, pageable);
                break;
            case "comments": // 댓글 좋아요
                ranking = userRepository.findRankingByCommentLikes(start, end, pageable);
                break;
            case "likes": // 토론 좋아요 (기본값)
            default:
                ranking = userRepository.findRankingByDebateLikes(start, end, pageable);
                break;
        }

        // 순위 부여
        for (int i = 0; i < ranking.size(); i++) {
            ranking.get(i).setRank((long) (i + 1));
        }

        return ranking;
    }

    private java.time.LocalDateTime calculateStartDate(String period, java.time.LocalDateTime end) {
        switch (period.toLowerCase()) {
            case "daily":
                return java.time.LocalDate.now().atStartOfDay();
            case "monthly":
                return java.time.LocalDate.now().withDayOfMonth(1).atStartOfDay();
            case "yearly":
                return java.time.LocalDate.now().withDayOfYear(1).atStartOfDay();
            case "all":
            default:
                return java.time.LocalDateTime.of(2000, 1, 1, 0, 0);
        }
    }

    /**
     * (Deprecated) 기존 메서드 유지 - 하위 호환성
     */
    public List<UserRankingResponse> getUserRanking(int limit) {
        return getUserRanking("all", "likes", limit);
    }
    public UserResponse getUserById(Long id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("사용자를 찾을 수 없습니다"));
        
        // 통계 정보 계산
        long debateCount = debateRepository.findByUserAndIsHiddenFalse(user, Pageable.unpaged()).getTotalElements();
        long commentCount = commentRepository.findByUser(user).size();
        
        // 받은 좋아요 수: 사용자가 작성한 토론들에 받은 좋아요 총합
        long likeCount = debateRepository.findByUserAndIsHiddenFalse(user, Pageable.unpaged())
                .getContent()
                .stream()
                .mapToLong(debate -> likeRepository.countByDebate(debate))
                .sum();
        
        // 참여한 토론 수: 입장을 선택한 토론 수
        long participatedCount = debateOpinionRepository.findByUser(user).size();
        
        return UserResponse.from(user, debateCount, commentCount, likeCount, participatedCount);
    }

    @Transactional
    public UserResponse updateProfile(Long userId, String nickname, String bio, String profileImage) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("사용자를 찾을 수 없습니다"));

        if (nickname != null && !nickname.isEmpty()) {
            user.setNickname(nickname);
        }
        if (bio != null) {
            user.setBio(bio);
        }
        if (profileImage != null) {
            user.setProfileImage(profileImage);
        }

        user = userRepository.save(user);
        return UserResponse.from(user);
    }


    public User getUserEntity(Long id) {
        return userRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("사용자를 찾을 수 없습니다"));
    }

}

File: DebateUser/DebateUserFrontEnd/src/services/userService.js
/**
 * 사용자 서비스 (User Service)
 *
 * 사용자 관련 API 호출을 담당하는 서비스입니다.
 *
 * 주요 기능:
 * - 사용자 정보 조회 (ID로)
 * - 프로필 수정
 */

import api from "./api";

export const userService = {
  /**
   * 사용자 정보 조회 (ID로)
   *
   * 사용자 ID로 사용자 정보를 가져옵니다.
   *
   * @param {number} id - 사용자 ID
   * @returns {Promise<Object>} ApiResponse 구조의 응답 데이터
   * @returns {Object} response.data - UserResponse (사용자 정보)
   */
  async getUserById(id) {
    const response = await api.get(`/users/${id}`);
    return response.data;
  },

  /**
   * 프로필 수정
   *
   * 현재 로그인한 사용자의 프로필을 수정합니다. 인증이 필요합니다.
   *
   * @param {string} nickname - 닉네임 (선택사항)
   * @param {string} bio - 자기소개 (선택사항)
   * @param {string} profileImage - 프로필 이미지 URL (선택사항)
   * @returns {Promise<Object>} ApiResponse 구조의 응답 데이터
   * @returns {Object} response.data - UserResponse (수정된 사용자 정보)
   */
  async updateProfile(nickname, bio, profileImage) {
    const response = await api.put("/users/me", null, {
      params: { nickname, bio, profileImage },
    });
    return response.data;
  },

  /**
   * 프로필 이미지 업로드
   * 
   * @param {File} file - 업로드할 이미지 파일
   * @returns {Promise<string>} 업로드된 이미지 URL
   */
  async uploadImage(file) {
    const formData = new FormData();
    formData.append('file', file);

    const response = await api.post('/upload/profile', formData);
    // ApiResponse<String> 구조에서 data 필드가 URL임
    return response.data.data || response.data;
  },

  /**
   * 특정 사용자가 작성한 토론 목록 조회
   * 
   * @param {number} userId - 사용자 ID
   * @param {number} page - 페이지 번호 (0부터 시작)
   * @param {number} size - 페이지당 항목 수
   * @returns {Promise<Object>} 페이징된 토론 목록
   */
  async getUserDebates(userId, page = 0, size = 10) {
    const response = await api.get(`/users/${userId}/debates`, {
      params: { page, size }
    });
    return response.data;
  },

  /**
   * 특정 사용자가 작성한 댓글 목록 조회
   * 
   * @param {number} userId - 사용자 ID
   * @param {number} page - 페이지 번호 (0부터 시작)
   * @param {number} size - 페이지당 항목 수
   * @returns {Promise<Object>} 페이징된 댓글 목록
   */
  async getUserComments(userId, page = 0, size = 10) {
    const response = await api.get(`/users/${userId}/comments`, {
      params: { page, size }
    });
    return response.data;
  },
};

/**
 * 사용자 랭킹 조회 (기간 및 기준별)
 * @param {number} limit - 조회할 상위 사용자 수
 * @param {string} period - 기간 (daily, monthly, yearly, all)
 * @param {string} criteria - 기준 (likes, votes, comments)
 * @returns {Promise} 사용자 랭킹 목록
 */
export const getUserRanking = async (limit = 10, period = 'all', criteria = 'likes') => {
  try {
    const response = await api.get("/users/ranking", {
      params: { limit, period, criteria },
    });
    return response.data;
  } catch (error) {
    console.error("랭킹 조회 실패:", error);
    throw error;
  }
};
